### PR TITLE
Support for data-parallelism for reduce, transform reduce, transform_binary_reduce algorithms

### PR DESCRIPTION
**Pull request by srinivasyadav18**
Contributes to Fixing #2333 

## Proposed Changes

  - Adapts new **vector_pack** trait `reduce` to support simd reductions.
  - Adapts new **CPO** `sequential_reduce` to support tag_fallback overloads on `reduce, transform_reduce, transform_reduce_binary` algorithms.
  - Adds tag dispatching overloads to support data-parallelism on `reduce, transform_reduce, transform_reduce_binary` algorithms.
  - Adds unit tests to newly supported data-parallel algorithms.

### CODE REVIEW

**Comment by hkaiser** [comment]
@srinivasyadav18 is this good to go now?

**Comment by srinivasyadav18** [comment]
@hkaiser Yes, it is ready now! Please review it. 
I will add perfect forwarding to all other CPOs used for datapar algorithms in #5858, because it will be easier to rebase.

File: libs/core/algorithms/include/hpx/parallel/algorithms/detail/reduce.hpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+        template <typename Iter1, typename Sent, typename Iter2, typename T,
+            typename Reduce, typename Convert>
+        friend inline constexpr T tag_fallback_invoke(sequential_reduce_t,
+            Iter1 first1, Sent last1, Iter2 first2, T init, Reduce&& r,
+            Convert&& conv)
+        {
+            util::loop2<ExPolicy>(first1, last1, first2,
+                [&init, reduce_op = HPX_FORWARD(Reduce, r),
+                    convert_op = HPX_FORWARD(Convert, conv)](
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/detail/reduce.hpp:92]: A similar construct above captures the function objects by reference, here you embedded them into the lambda. Both solutions are probably ok (if we're sure the arguments don't go out of scope), I'd suggest however to stay consistent.

File: libs/core/algorithms/include/hpx/parallel/algorithms/transform_reduce.hpp [hunk 2]
```diff
@@ -525,52 +391,12 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
-                                HPX_FORWARD(Op2, op2), result});
-                    }
-
-                    return util::detail::extract_value<ExPolicy>(result);
+                    auto result = HPX_INVOKE(op2, *it1, *it2);
+                    ++it1;
+                    ++it2;
+
+                    return detail::sequential_reduce<ExPolicy>(it1, last1, it2,
+                        result, HPX_FORWARD(Op1, op1), HPX_FORWARD(Op2, op2));
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/transform_reduce.hpp:399]:
```suggestion
                    auto&& result = HPX_INVOKE(op2, *it1, *it2);
                    ++it1;
                    ++it2;

                    return detail::sequential_reduce<ExPolicy>(it1, last1, it2,
                        HPX_MOVE(result), HPX_FORWARD(Op1, op1), HPX_FORWARD(Op2, op2));
```

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

**Comment by hkaiser** [comment]
bors merge

**Pull request merged**